### DISCUSSION TITLE
ORM / sqlite3 - using session to loop over one row at a time (loaded as ORM object) seems to accumulate results in memory

### DISCUSSION THREAD

**Discussion by jeanmonet** [sqlalchemy/sqlalchemy#8198] [Usage Questions]
### Describe the bug

Using the ORM paradigm, looping over each row, seems to lead to an accumulation in memory of all previously loaded objects, even if the python variable is overwritten on each loop.

Switching to Core, the problem disappears.

### To Reproduce

```python
# ------- ORM approach (NOT OK) leading to an accumulation of objects in memory ---------
class MySqlAlchemyORMModel(...):
    ...
    id: Column(Integer, primary_key=True)
    ...

engine = ...
session = ...
for row in session.execute( select(MySqlAlchemyORMModel) ):
    # `row` is an instance of `MySqlAlchemyORMModel`
    print(row.id)
    # It appears that after each loop, the previously loaded object remains somewhere in memory
    # even though the `row` variable is overwritten.
    # This leads to running out of memory eventually, if the table size is larger than RAM.

# -------- Core approach (OK) ----------

engine = ...

with engine.connect() as conn:
    for row in conn.execute( sql.select(MySqlAlchemyORMModel) ):
        # `row` is a tuple
        print(row[0])
        # No memory accumulation effect
```

### Error

```
# running `htop` shows how memory is being accumulated by the process.
``` 

### Versions

- OS: Ubuntu 20.04
- Python: 3.10
- SQLAlchemy: 1.4.39
- Database: SQLite
- DBAPI (eg: psycopg, cx_oracle, mysqlclient):  sqlite 3.39.0

**Comment by zzzeek** [comment]
hi -

there's not enough actionable information for us to help with this issue, and it could just as well be a usage-related effect, but it's impossible to say without any detail.   

we would need to see an exact runnable [MCVE](http://stackoverflow.com/help/mcve) script (see link for requirements) that shows the memory growth for us to begin analyzing what might be going on.

note that the pysqlite driver itself has an unusual memory behavior that uses internal caching and will appear to grow for some time, however it is self-limiting.

**Comment by jeanmonet** [comment]
Hi, thanks for the quick reply. I have included an fully illustrated example below evidencing the behavior.

I must say I spent countless hours trying to debug this before realizing it is SQLAlchemy-related (at least when used with SQLite - I haven't tested with Postgres and didn't install any specific db driver for sqlite, just `sqlite3`).

```bash
$ conda list |grep sql
sqlalchemy                1.4.39          py310h5764c6d_0    conda-forge
sqlite                    3.39.0               h4ff8645_0    conda-forge
```

The memory accumulation is quite severe however, it quickly captures all of my RAM when working with lengthy data.

```python
from pathlib import Path
import psutil

import sqlalchemy as sa
from sqlalchemy.orm import sessionmaker
from sqlalchemy.orm import scoped_session
from sqlalchemy.ext.declarative import declarative_base

# ---- Parameters ----
db_file = Path("/tmp/testdb.sqlite")
n_gigas = 3    # Approximate number of GigaBytes to write to db_file
# --------------------

db_file.unlink(missing_ok=True)        # start from empty db file

engine = sa.create_engine(f"sqlite:///{db_file!s}")
Session = sessionmaker(bind=engine)
Base = declarative_base()

class MyTestModel(Base):
    __tablename__ = "mytestmodel"
    id = sa.Column(sa.Integer, primary_key=True)
    data = sa.Column(sa.Text)

Base.metadata.create_all(engine)

# Populate database
with Session() as session:
    print(f"Writing approx {n_gigas}GB to db file:", str(db_file))
    print(f"[Before] db file size is {round(db_file.stat().st_size / 1024**2, 1)}MB")
    for i in range(n_gigas):
        # Writing approx 1GB per loop
        session.add_all(
            (MyTestModel(data="a" * 10 * 1024**2) for _ in range(100))  # 100 row 10MB each = 1000MB
        )
    session.commit()
    print(f"[After] db file size is {round(db_file.stat().st_size / 1024**2, 1)}MB")

mem_before = round(psutil.virtual_memory().used / 1024**3, 1)
print(f"Memory usage before iterating on rows: {mem_before}GB")
with Session() as session:
    for (row,) in session.execute(sa.select(MyTestModel)):
        pass
        # print(row.id, end=" ")
mem_after = round(psutil.virtual_memory().used / 1024**3, 1)
print(f"Memory usage after iterating on rows: {mem_after}GB.")
print(f"Diff={mem_after - mem_before:.1f}GB, which is approximately equal to {n_gigas}GB.")
print("This illustrates a memory accumulation when iterating over rows using ORM pattern.")
```

Output on my machine and with above settings:

```
Writing approx 3GB to db file: /tmp/testdb.sqlite
[Before] db file size is 0.0MB
[After] db file size is 3003.5MB
Memory usage before iterating on rows: 9.6GB
Memory usage after iterating on rows: 12.6GB.
Diff=3.0GB, which is approximately equal to 3GB (db size).
This illustrates a memory accumulation when iterating over rows using ORM pattern.
```

**Reply by zzzeek** [reply]
Thanks for the clear test script illustrating your methodologies.   

So when the Session loads ORM objects, without additional steps (see below) it buffers them all at once before yielding them.  this is in contrast to working with Core where we deliver Row objects directly from the DBAPI cursor and there's no buffering on the SQLAlchemy side.  Database drivers will often buffer rows before returning them from the cursor, but in the case of the sqlite3 driver, this does not seem to be the case; if we read one row at a time, it only allocates a relatively small amount of memory, seems here to be a row at a time if not just a few.

So here is the baseline run with your script on my machine - the memory growth is from 3.4 to 5.7 G using the ORM example you gave:

```
Writing approx 3GB to db file: /tmp/testdb.sqlite
[Before] db file size is 0.0MB
[After] db file size is 3003.5MB
Memory usage before iterating on rows: 3.4GB
size: None
Memory usage after iterating on rows: 5.7GB.
Diff=2.3GB.
```

let's take the ORM select out entirely, and even Core.  since the majority of memory here is the very huge rows you have, let's see what plain old sqlite3 does, *if we don't discard each row as we read it*; also just for a moment we will not clear out the rows, will address that next:

```py
import sqlite3
sqlite_connection = sqlite3.connect("/tmp/testdb.sqlite")
cursor = sqlite_connection.cursor()
cursor.execute("select * from mytestmodel")
rows = cursor.fetchall()
cursor.close()

# not deleting the rows yet
# del rows

mem_after = round(psutil.virtual_memory().used / 1024**3, 1)
print(f"Memory usage after iterating on rows: {mem_after}GB.")
print(f"Diff={mem_after - mem_before:.1f}GB")
```

for that we get output:

```
Writing approx 3GB to db file: /tmp/testdb.sqlite
[Before] db file size is 0.0MB
[After] db file size is 3003.5MB
Memory usage before iterating on rows: 3.5GB
Memory usage after iterating on rows: 5.7GB.
Diff=2.2GB.

```

output is the same, memory growth is 3.3G to 5.7G.   So the ORM doesn't have any unusual memory use here.  What's different is if we take the above script and delete that set of rows before measuring memory:

```py
del rows
```

then the net memory at the end doesn't grow at all, because it was cleared:

```
Writing approx 3GB to db file: /tmp/testdb.sqlite
[Before] db file size is 0.0MB
[After] db file size is 3003.5MB
Memory usage before iterating on rows: 3.4GB
Memory usage after iterating on rows: 3.5GB.
Diff=0.1GB 
```


OK, so we are closing out `Session()` why isn't the memory cleared?  This has to do with the nature of the ORM objects being more complex structures, so are not cleared unless we `del` the two objects explicitly, which happens normally if we put the work we're doing inside of a function, so is not too much different from the sqlite3 example:

```py
with Session() as session:
    for (row, ) in session.execute(sa.select(MyTestModel)):
        pass

mem_after = round(psutil.virtual_memory().used / 1024**3, 1)
print(f"Memory usage after iterating on rows: {mem_after}GB.")
print(f"Diff={mem_after - mem_before:.1f}GB ")

del row
del session

mem_after = round(psutil.virtual_memory().used / 1024**3, 1)
print(f"Memory usage after del: {mem_after}GB.")
print(f"Diff={mem_after - mem_before:.1f}GB ")

```

now we see the memory cleared (note the last memory usage line "After del").  the .2 G is 100M more than .1G, we can likely chalk this up to some internal ORM result fetching structures that are cached being a bit more complex:

```
Writing approx 3GB to db file: /tmp/testdb.sqlite
[Before] db file size is 0.0MB
[After] db file size is 3003.5MB
Memory usage before iterating on rows: 3.5GB
Memory usage after iterating on rows: 5.9GB.
Diff=2.4GB 
Memory usage after GC: 3.7GB.
Diff=0.2GB 

```

So we can see there's nothing unusual happening here in the ORM at all except that a giant result set will be buffered all at once before yielding it, then there's a little bit of Python stickiness but nothing too intense.


Now why does the ORM buffer all the objects?  It's because the ORM has a heavy emphasis on the two concepts of [the identity map](https://docs.sqlalchemy.org/en/14/glossary.html#term-identity-map) as well as [eager loading](https://docs.sqlalchemy.org/en/14/glossary.html#term-eager-loading).  Both of these concepts work much better when **aggregated over many objects at once**; if we kept throwing away rows as they were read, some eager loading strategies might not work well with that, and if you have a result set that might have the same object identity expressed multiple times (which is common with eager loading as well as some kinds of joins), if the object itself kept changing as the result were read, that can really mess up *some* kinds of code that might be working with those objects.

So when you load a huge ORM result set, you want to use **partitioning** of some kind.    The most common partitioning people do is [SQL-level pagination](https://use-the-index-luke.com/blog/2013-07/pagination-done-the-postgresql-way); this is popular because often, people are doing some kind of stateless transfer or similar.  We have some examples using window functions at https://github.com/sqlalchemy/sqlalchemy/wiki/RangeQuery-and-WindowedRangeQuery for this kind of thing.

In this case however there's a much easier solution which is to use [yield_per()](https://docs.sqlalchemy.org/en/14/orm/queryguide.html#yield-per) (I'm also updating some docs on this right now as my review of that link illustrated some missing detail).  This instructs the ORM to only build up a certain number of ORM objects before yielding them out, and is what you want to do in this case.  optionally combined with [result.partitions()](https://docs.sqlalchemy.org/en/14/core/connections.html?highlight=result+partitions#sqlalchemy.engine.Result.partitions) it looks like this:

```py
with Session() as session:
    for partition in session.execute(sa.select(MyTestModel)).yield_per(10).partitions():
        for (row,) in partition:
            pass

mem_after = round(psutil.virtual_memory().used / 1024**3, 1)
print(f"Memory usage after iterating on rows: {mem_after}GB.")
print(f"Diff={mem_after - mem_before:.1f}GB ")

del partition
del row
del session

mem_after = round(psutil.virtual_memory().used / 1024**3, 1)
print(f"Memory usage after del: {mem_after}GB.")
print(f"Diff={mem_after - mem_before:.1f}GB ")
```

running that, we see that only 10 ORM objects are in memory at once and we then don't see excessive memory use at once:

```
Writing approx 3GB to db file: /tmp/testdb.sqlite
[Before] db file size is 0.0MB
[After] db file size is 3003.5MB
Memory usage before iterating on rows: 3.5GB
Memory usage after iterating on rows: 3.9GB.
Diff=0.4GB 
Memory usage after del: 3.7GB.
Diff=0.2GB 
```


`yield_per()` can be used by itself without `partitions()` alone so i think you likely want to use that.

**Reply by jeanmonet** [reply]
First of all, thank you very much for the detailed and in-depth response. It is immensely helpful for better understanding the behind-the-scenes apparatus of the ORM and its interaction with the database driver.

I am today years old since I've learned about `.yield_per()` and the huge difference it makes (even if I've been a SQLAlchemy fan for a few years now). In fact, without it I was simply unable to make full use of the ORM because of those memory issues.

> Database drivers will often buffer rows before returning them from the cursor, but in the case of the sqlite3 driver, this does not seem to be the case; if we read one row at a time, it only allocates a relatively small amount of memory, seems here to be a row at a time if not just a few.
> ... So we can see there's nothing unusual happening here in the ORM at all except that a giant result set will be buffered all at once before yielding it, then there's a little bit of Python stickiness but nothing too intense.

Regarding this point, I was curious and sampled the memory usage every 10 rows during the iteration (*without using `.yield_per()`*). It appears that the full amount of memory allocation is incurred upfront, starting with the first sample. So memory allocation did not increase progressively as the for loop progresses, but upfront at the start of the for loop, and remains constant until the end. Thus I assume the results were entirely buffered upfront by the ORM itself, and I should expect the same behavior with other databases such as Postgres.

```python
...
mem_before = round(psutil.virtual_memory().used / 1024**3, 1)
print(f"Memory usage before iterating on rows: {mem_before}GB")
mem_in_time = [mem_before]
with Session() as session:
    for i, (row,) in enumerate(
            session.execute(sa.select(MyTestModel))     # .yield_per(1)
    ):
        # ------- MEMORY SNAPSHOT AT REGULAR INTERVALS -------
        if i % 10:
            # Take snapshot of memory every 10 rows
            mem_in_time.append(round(psutil.virtual_memory().used / 1024**3, 1))
        # ---------------------------------------------------------------
        # print(row.id, end=" ")
mem_after = round(psutil.virtual_memory().used / 1024**3, 1)
mem_in_time.append(mem_after)
print(f"Memory usage after iterating on rows: {mem_after}GB.")

from matplotlib import pyplot as plt
_ = plt.plot(mem_in_time)
_ = plt.ylabel("Mem in GB")
# if not running in jupyter:  plt.show()
```

**Reply by zzzeek** [reply]
Just an FYI from this issue I found the existing yield_per API and docs to be problematic in many regards, including that I forgot to add it to asyncio and also using Core-only was too confusingly different.  so in 1.4.40 a large improvement to these APIs are dropping and I've also rewritten all the docs, code is in review at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/3964

**Reply by zzzeek** [reply]
but yes the ORM normally buffers the whole result and all the objects it's giong to return.  because it has to apply eager loaders before yielding them.

**Reply by jeanmonet** [reply]
Noted, thanks for the heads-up.
I was looking at the proposed changes in the query guide https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/3963/10/doc/build/orm/queryguide.rst#1097, that I find helpful. Just a small note, there may be a typo on that line at *selectinload*?
> potentially compatible with selectinload eager loading, provided the database

**Reply by zzzeek** [reply]
it's referring to https://docs.sqlalchemy.org/en/14/orm/loading_relationships.html?highlight=selectinload#sqlalchemy.orm.selectinload .  i can make that more of a real doc link for context later

**Reply by jeanmonet** [reply]
Oh makes sense now! Indeed a doc link would probably help.
